blueflood-core/src/test/java/com/rackspacecloud/blueflood/utils/ModuleLoaderTest.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    @Test
+    public void givingDifferentKeysButSameClassYieldsTwoSeparateObjects() {
+
+        // given
+        Configuration.getInstance().setProperty(
+                CoreConfig.DISCOVERY_MODULES,
+                "com.rackspacecloud.blueflood.io.DummyDiscoveryIO");
+        Configuration.getInstance().setProperty(
+                CoreConfig.ENUMS_DISCOVERY_MODULES,
+                "com.rackspacecloud.blueflood.io.DummyDiscoveryIO");
+
+        // when
+        Object loadedModule1 = ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.DISCOVERY_MODULES);
+
+        // then
+        Assert.assertNotNull(loadedModule1);
+        Assert.assertEquals(DummyDiscoveryIO.class, loadedModule1.getClass());
+
+        // when
+        Object loadedModule2 = ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.ENUMS_DISCOVERY_MODULES);
+
+        // then
+        Assert.assertNotNull(loadedModule2);
+        Assert.assertEquals(DummyDiscoveryIO.class, loadedModule1.getClass());
+        Assert.assertNotSame(loadedModule1, loadedModule2);
+    }
+
+    @Test
+    public void callingTwiceWithSameKeyButDifferentClassesYieldsTheSameObjectBothTimes() {
+
+        // TODO: This functionality is extremely misleading. It's probably not
+        // what was originally intended, anyways. Basically, if we pass the
+        // same CoreConfig enum field a second time to getInstance after
+        // changing the configuration value for that key, ModuleLoader will
+        // still give us back the object that it loaded the first time. This is
+        // the case even if we specify completely incompatible classes each
+        // time.
+
+        // given
+        Configuration.getInstance().setProperty(
+                CoreConfig.DISCOVERY_MODULES,
+                "com.rackspacecloud.blueflood.io.DummyDiscoveryIO");
+
+        Object loadedModule1 = ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.DISCOVERY_MODULES);
+
+        // precondition
+        Assert.assertNotNull(loadedModule1);
+        Assert.assertEquals(DummyDiscoveryIO.class, loadedModule1.getClass());
+
+        // when
+        Configuration.getInstance().setProperty(
+                CoreConfig.DISCOVERY_MODULES,
+                "com.rackspacecloud.blueflood.io.DummyDiscoveryIO2");
+        Object loadedModule2 = ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.DISCOVERY_MODULES);
+
+        // then
+        Assert.assertNotNull(loadedModule2);
+        Assert.assertSame(loadedModule1, loadedModule2);
+    }
+
+    @Test
+    public void specifyingAnInterfaceYieldsNull() {
+
+        // given
+        Configuration.getInstance().setProperty(
+                CoreConfig.DISCOVERY_MODULES,
+                "com.rackspacecloud.blueflood.io.DiscoveryIO");
+
+        // when
+        Object loadedModule = ModuleLoader.getInstance(DiscoveryIO.class, CoreConfig.DISCOVERY_MODULES);
+
+        // then
+        Assert.assertNull(loadedModule);
+    }
+}
```

Comment:
Also for a class which is package-protected and is from a different package than the test case.